pubsub/gochannel/fanin.go

```diff
@@ -0,0 +1,139 @@
+package gochannel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+// FanIn is a component that receives messages from 1..N topics from a subscriber and publishes them
+// on a specified topic. In effect, messages are "multiplexed".
+//
+// You need to call AddSubscription method with a slice of topics that you want to
+// listen to and a topic you want them to be published to.
+// This needs to be done *before* starting the FanIn.
+//
+// FanIn exposes the standard Subscriber interface.
+type FanIn struct {
+	internalPubSub *GoChannel
+	internalRouter *message.Router
+
+	subscriber message.Subscriber
+
+	logger watermill.LoggerAdapter
+
+	subscriptions  map[string]struct{}
+	subscribedLock sync.Mutex
+}
+
+// NewFanIn creates a new FanIn.
+func NewFanIn(
+	subscriber message.Subscriber,
+	logger watermill.LoggerAdapter,
+) (*FanIn, error) {
+	if subscriber == nil {
+		return nil, errors.New("missing subscriber")
+	}
+	if logger == nil {
+		logger = watermill.NopLogger{}
+	}
+
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &FanIn{
+		internalPubSub: NewGoChannel(Config{}, logger),
+		internalRouter: router,
+		logger:         logger,
+		subscriber:     subscriber,
+		subscriptions:  map[string]struct{}{},
+	}, nil
+}
+
+// AddSubscription adds an internal subscriptions.
+// You need to call this method with slice of `fromTopics` and `toTopic` before the FanIn is started.
+// AddSubscription is idempotent.
+func (f *FanIn) AddSubscription(fromTopics []string, toTopic string) {
+	f.subscribedLock.Lock()
+	defer f.subscribedLock.Unlock()
+
+	if len(fromTopics) == 0 {
+		return
+	}
+
+	key := createKey(fromTopics, toTopic)
+	_, ok := f.subscriptions[key]
+	if ok {
+		// Subscription already exists
+		return
+	}
+
+	f.logger.Trace("Adding fan-in subscription for topics", watermill.LogFields{
+		"fromTopics": fromTopics,
+		"toTopic":    toTopic,
+	})
+
+	for _, fromTopic := range fromTopics {
+		f.internalRouter.AddHandler(
+			fmt.Sprintf("fanin-from-%s-to-%s", fromTopic, toTopic),
+			fromTopic,
+			f.subscriber,
+			toTopic,
+			f.internalPubSub,
+			message.PassthroughHandler,
+		)
+	}
+
+	f.subscriptions[key] = struct{}{}
+}
+
+// Run runs the FanIn.
+func (f *FanIn) Run(ctx context.Context) error {
+	return f.internalRouter.Run(ctx)
+}
+
+// Running is closed when FanIn is running.
+func (f *FanIn) Running() chan struct{} {
+	return f.internalRouter.Running()
+}
+
+// Subscribe implements the standard Subscriber interface.
+func (f *FanIn) Subscribe(ctx context.Context, topic string) (<-chan *message.Message, error) {
```

Comment:
Does this method still make sense for `FanIn`? It seems unused now.

Reply:
this also came to my mind, how it should be used :thinking:

Reply:
Ah oh, I got it.

Maybe it would be better to not expose subscriber here, but just provide higher level API that is doing that under the hood? In other words - to be able to use FanIn without creating the router. So in that case I could just do

```
fanin, err := gochannel.NewFanIn(upstreamPubSub, config, logger)
// ...
fanin.Run()
```

In that case API may be much simpler to use. But the question is if in that case it fits to your use case ;-)